File: LeetCode/Apple/TopQuestions/Dynamic_Programming/LongestPalindrome.py
'''
Greedy Approach
'''
class Solution:
    def longestPalindrome(self,s:str)-> str:
        left = 0#0
        right = 2#2
        #save start index in different varaible
        #not update
        indices = []
        for i in range(1,len(s)-1):

            while left >= 0 and right < len(s):
                if s[left]==s[right]:
                    indices.append(left) #first left
                    left -=1
                    right +=1
            indices.append(right)
        return s[indices[0]:indices[-1]]

class Solution(object):
    def longestPalindrome(self, s):
        largestPalindrome = ""
        for i in range(len(s)):
            palindromeOdd = self.largestPalindromeAtIndex(s, i, i)
            palindromeEven = self.largestPalindromeAtIndex(s, i, i + 1)
            largerPalindrome = palindromeOdd if len(palindromeOdd) > len(palindromeEven) else palindromeEven
            largestPalindrome = largestPalindrome if len(largestPalindrome) >= len(largerPalindrome) else largerPalindrome
        return largestPalindrome

    def largestPalindromeAtIndex(self, s, left, right):
        leftIndex = 0
        rightIndex = 0
        while left >= 0 and right < len(s):
            if s[left] == s[right]:
                leftIndex = left
                rightIndex = right
            else:
                break
            left -= 1
            right += 1
        return s[leftIndex:rightIndex+1]

s = Solution()
print (s.longestPalindrome('babaabbaa'))


if __name__ == '__main__':
    #s = "ababa"
    s = "xaaabbaa"
    output = Solution()
    print ('The Longest Palindrome is:',output.longestPalindrome(s))

File: LeetCode/Apple/TopQuestions/Dynamic_Programming/LongestPalindrome.cpp
//
//Dynamic Programming Solution in C++
#include <iostream>
#include <map>
#include <string>
#include <algorithm>
#include <vector>

using namespace std;
class Solution{
public:
  string longestPalindrome(string & s)
  {
    int left = 0;
    int right = 2;
    vector <int> indices={};
    for (int i=1;i<s.length()-1;i++)
    {
      while (left>=0 && right < s.length())
      {
        if (s[left]==s[right])
        {
          indices.push_back(left);
          left -=1;
          right +=1;
        }
      }
      indices.push_back(right);
    }
    return s.substr(indices[0],indices[-1]);
  }
};

int main()
{
  string s = "ababa";
  Solution palindrome;
  cout << palindrome.longestPalindrome(s) << endl;
  return 0;
}
